text/0000-parallel-model-hooks.md

```diff
@@ -0,0 +1,115 @@
+This has two major implications:
+
+1. The `afterModel` hook will be passed a promise, since the result of the `model` hook will not have resolved by the time `afterModel` is called.
+2. Routes may not have resolved the data requested by `modelFor` when it is called. It will need to return a promise or there will need to be another method for getting a promise that eventually resolves a route's model. (`modelPromiseFor`?)
+
+#### Router.js (tildeio/router.js#171)
+
+`handler-info` presently has a `resolve` method that calls the model hooks. We move the model hook calls to a new `request` method, and save the promises on the `handler-info`'s context (the promise from the `model` hook is passed to the `afterModel` hook). The `resolve` method takes the stored promises and inserts them into the resolution chain.
+
+`transition-state` presently has a `resolveOneHandlerInfo` method that recurses through the `handler-info`s for a transition and calls the associated `resolve` method. Before this, we insert a `requestOneHandlerInfo` method that recurses through `handler-info`s, calling each one's `request` method, which now executes the model hooks, kicking off any network requests.
+
+We add a `requestIndex` property to `transition` that keeps track of how many route's model hooks we have already run (akin to how `resolveIndex` presently keeps track of how many routes we have resolved).
+
+#### Ember.js (emberjs/ember.js#12415)
+
+The default model hook uses `requestIndex` (instead of `resolveIndex`) to grab `modelPromise` (instead of `context`) from the `handler-info`.
+
+`Route#modelFor` returns a promise.
+
+#### Usage
+
+Routes that don't depend on another route's data likely don't need any changes, but their requests are sent in parallel:
+
+```javascript
+// app/routes/parent.js
+export default Ember.Route.extend({
+  model() {
+    return Ember.$.get('/api/foo');
+  }
+});
+
+// app/routes/parent/child.js
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  }
+});
+```
+
+Routes that depend on another route's data must consume that data as a promise:
+
+```javascript
+// app/routes/parent.js
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  }
+});
+
+// app/routes/parent/child.js
+export default Ember.Route.extend({
+  model() {
+    return this.modelFor('parent').then((parentModel) => {
+      return Ember.$.get(`/api/bar/${parentModel.get('baz')}`);
+    });
+  }
+});
+```
+
+Routes whose `afterModel` hook operates on the resolved model must now consume a promise:
+
+```javascript
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  },
+  afterModel(modelPromise, transition) {
+    return modelPromise.then((data) => {
+      if (data.abort) {
+        this.transitionTo('fiz');
+      }
+    }
+  }
+});
+```
+
+# Drawbacks
+
+- This is a major breaking change.
```

Comment:
@rwjblue yup.

I think a separate name, and provided via community add-on (with minimal hooks in internals) may be an escape valve. I want us to be careful not to cram ideas (good or bad) too quickly into ember. As once they are added, it is tricky and extremely costly to unwind.

We can look to several hasty (although well thought on 1.13.x features) as a demonstration of this.

I do believe the motivation here is good, the implementation may also. But it is well known the more users we can get using this in the wild the faster we can explore the problem space, and the more unknown unknowns we can squash.